src/io/linux.zig

```diff
@@ -215,9 +300,128 @@ pub const IO = struct {
         };
         completion.prep(sqe);
 
+        if (self.awaiting) |first| {
+            assert(self.awaiting_count > 0);
+            assert(first.awaiting_previous == null);
+            first.awaiting_previous = completion;
+            completion.awaiting_next = first;
+        } else {
+            assert(self.awaiting_count == 0);
+        }
+        self.awaiting = completion;
+        self.awaiting_count += 1;
+
         self.ios_queued += 1;
     }
 
+    /// Cancel should be invoked at most once, before any of the memory owned by read/recv buffers
+    /// is freed (so that lingering async operations do not write to them).
+    ///
+    /// After this function is invoked:
+    /// - No more completion callbacks will be called.
+    /// - No more IO may be submitted.
+    ///
+    /// This function doesn't return until either:
+    /// - All events submitted to io_uring have completed.
+    ///   (They may complete with `error.Canceled`).
+    /// - Or, an io_uring error occurs.
+    ///
+    /// TODO(Linux):
+    /// - Linux kernel ≥5.19 supports the IORING_ASYNC_CANCEL_ALL and IORING_ASYNC_CANCEL_ANY flags,
+    ///   which would allow all events to be cancelled simultaneously with a single "cancel"
+    ///   operation, without IO needing to maintain the `awaiting` doubly-linked list and the `next`
+    ///   cancellation stage.
+    /// - Linux kernel ≥6.0 supports `io_uring_register_sync_cancel` which would remove the `queued`
+    ///   cancellation stage.
+    pub fn cancel(self: *IO) !void {
```

Comment:
Would it make sense to have this implicitly in `deinit()` instead? Need to change the test case if so.

If not, think `main.zig`/`repl.zig`/other uses of `IO.init` need to cancel as well. Or at least assert `awaiting_count == 0` in `IO.deinit()`.

Reply:
> Would it make sense to have this implicitly in `deinit()` instead? Need to change the test case if so.

By the time `io.deinit()` is being called, the buffers have already been freed, so it is too late.

> If not, think main.zig/repl.zig/other uses of IO.init need to cancel as well. Or at least assert awaiting_count == 0 in IO.deinit().

Since `main.zig` and `repl.zig` only exit when the process exits, I don't think we actually need to cancel IO in those cases.

Reply:
Ah, the buffer lifetimes are beginning to click. Thanks.